### ISSUE TITLE
SIGSEGV Causing spire-server Container to Crash During Calls to GetFederatedBundle Endpoint

### ISSUE THREAD

**Issue by dcortesantonio** [spiffe/spire#5005]
**Description:**
After introducing a new change in our entries, we've observed an increase in calls to the GetFederatedBundle endpoint and sometimes we got a connection issue to the DB cluster. Unfortunately, this has led to intermittent crashes of the spire-server container due to a SIGSEGV signal. 

<details>
<summary>Logs</summary>

```
"authorized_as":"nobody","authorized_via":"","caller_addr":"127.0.0.1:<redacted>","caller_id":"spiffe://<redacted>","error":"datastore-sql: dial tcp: lookup <redacted>-db.cluster.us-east-1.rds.amazonaws.com: Try again","level":"error","method":"GetFederatedBundle","msg":"Failed to fetch caller entries","request_id":"df375bf0-ea74-4fb6-ab44-3ef61a1a7c6a","service":"bundle.v1.Bundle","subsystem_name":"api","time":"2024-03-19T15:27:40Z"}
{"authorized_as":"nobody","authorized_via":"","caller_addr":"127.0.0.1:<redacted>","caller_id":"spiffe://<redacted>","error":"datastore-sql: dial tcp: lookup <redacted>-db.cluster.us-east-1.rds.amazonaws.com: Try again","level":"error","method":"GetAuthorizedEntries","msg":"Unable to look up agent information","request_id":"3cba7cdd-0bf4-49a4-ad48-4cd63879c165","service":"entry.v1.Entry","subsystem_name":"api","time":"2024-03-19T15:27:40Z"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x3009f7e]

goroutine 199442 [running]:
github.com/spiffe/spire/pkg/server/api/rpccontext.Logger({0x0?, 0x0?})
        /spire/pkg/server/api/rpccontext/alias.go:16 +0x1e
github.com/spiffe/spire/pkg/server/api/middleware.(*authorizationMiddleware).Preprocess(0x3bc6ad2?, {0x4159b58?, 0xc000da3140?}, {0x3bc6aaf, 0x35}, {0x385c4e0, 0xc02ad7f140})
        /spire/pkg/server/api/middleware/authorization.go:63 +0x3e5
github.com/spiffe/spire/pkg/common/api/middleware.middlewares.Preprocess({0xc0001a0020, 0x2, 0x2}, {0x4159b58?, 0xc000da3140?}, {0x3bc6aaf, 0x35}, {0x385c4e0, 0xc02ad7f140})
        /spire/pkg/common/api/middleware/middleware.go:87 +0xb6
github.com/spiffe/spire/pkg/common/api/middleware.middlewares.Preprocess({0xc0001a0010, 0x3, 0x3}, {0x4159b58?, 0xc000da3110?}, {0x3bc6aaf, 0x35}, {0x385c4e0, 0xc02ad7f140})
        /spire/pkg/common/api/middleware/middleware.go:92 +0x132
github.com/spiffe/spire/pkg/common/api/middleware.middlewares.Preprocess({0xc0001a0000, 0x4, 0x4}, {0x4159b58?, 0xc000da2f60?}, {0x3bc6aaf, 0x35}, {0x385c4e0, 0xc02ad7f140})
        /spire/pkg/common/api/middleware/middleware.go:92 +0x132
github.com/spiffe/spire/pkg/common/api/middleware.UnaryInterceptor.func1({0x4159b58?, 0xc000da2f60?}, {0x385c4e0, 0xc02ad7f140}, 0xc01dbd6780, 0xc01830ff68)
        /spire/pkg/common/api/middleware/interceptor.go:15 +0x97
github.com/spiffe/spire-api-sdk/proto/spire/api/server/bundle/v1._Bundle_GetFederatedBundle_Handler({0x39eb9a0?, 0xc000664540}, {0x4159b58, 0xc000da2f60}, 0xc009bf1a40, 0xc00030f170)
        /go/pkg/mod/github.com/spiffe/spire-api-sdk@v1.2.5-0.20221020001527-5895a0279944/proto/spire/api/server/bundle/v1/bundle_grpc.pb.go:376 +0x138
google.golang.org/grpc.(*Server).processUnaryRPC(0xc007ca2000, {0x41640c0, 0xc0025849c0}, 0xc018326480, 0xc007c9a4b0, 0x59789b8, 0x0)
        /go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:1340 +0xd23
google.golang.org/grpc.(*Server).handleStream(0xc007ca2000, {0x41640c0, 0xc0025849c0}, 0xc018326480, 0x0)
        /go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:1713 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        /go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:963 +0x28a
```
</details>

Upon investigation, we discovered that in one error step, the code is returning a nil context instead of the expected context object (as seen [here](https://github.com/spiffe/spire/blob/release/v1.5.5/pkg/server/api/middleware/entries.go#L46)). Consequently, when attempting to use the context in subsequent operations ([here](https://github.com/spiffe/spire/blob/release/v1.5.5/pkg/server/api/middleware/authorization.go#L63)), a nil pointer error occurs.

**Impact:**
The current issue is causing disruptions in the operation of the spire-server container, resulting in potential downtime and increased operational overhead due to the need to recreate connections.

* **Version**: v1.5.5
* **Platform**: 
* **Subsystem**:  Spire server

**Comment by faisal-memon** [comment]
hmm, should we be just returning ctx instead if nil on error?

**Comment by dcortesantonio** [comment]
> hmm, should we be just returning ctx instead if nil on error?

Yes, exactly.

**Comment by evan2645** [comment]
@dcortesantonio any chance you're up for contributing the fix?

For whomever is able to pick it up, it would be good to do a quick survey and see if there are any other spots like this , and put a couple tests around it 🙏

**Comment by amartinezfayo** [comment]
This issue has already been addressed in 1.6.0, please refer to #3829.
Could you please update to 1.6.0 or later and confirm that this has been solved for you?
I'm closing this issue for now, please re-open if needed.
